### ISSUE TITLE
`uninhabited_reference` is very likely to be a false positive

### ISSUE THREAD

**Issue by apoelstra** [rust-lang/rust-clippy#11984]
### Summary

The new `uninhabited_reference` lint complains about deferencing a reference to an infallible type, saying this is UB. But it is possible to write such code in safe Rust, which will be provably impossible to call (absent unsafe code which itself is almost certainly incorrect).

The lint purpose would be better served by linting the *creation* of a reference to an uninhabited type. The original issue has some discussion saying that it's maybe not UB, but it's almost certainly wrong. Versus the dereference, which is safe to write and has legitimate usecases.

At the very least, the existing lint shouldn't trigger on trait implementations.

cc #11851 

### Reproducer

```rust
impl PushBytesErrorReport for core::convert::Infallible {
    #[inline]
    fn input_len(&self) -> usize { match *self {} }
}
```

triggers the lint. But this code is clearly implementing an error-related trait for `Infallible`, with the understanding that the trait method is impossible to call.

**Comment by Kixunil** [comment]
Also a different perspective: since `unsafe` code could create the reference then by the logic of this lint **all** dereferences should be warned against because they could be UB...

**Comment by kpreid** [comment]
> The lint purpose would be better served by linting the creation of a reference to an uninhabited type.

[T-opsem is debating this point at length.](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Matching.20on.20references.20to.20uninhabted.20types)

(I agree that this lint should not be firing on `match *my_ref {}`; that's what I came here to report before I found it was already reported.)

**Comment by TethysSvensson** [comment]
I agree that this is a false positive. The lint should be disabled, or at least permit code in the form `fn method(..., value: &Empty, ...) { match *my_ref {} }` until the debate about operational semantics have been settled.

This pattern is used in a lot of places and as far as I know it is the idiomatic way of telling the compiler "this function cannot be called and all references to it should be optimized away". I do not know of any alternatives to this syntax in safe rust.

This pattern is especially useful when you have a lot of generic trait impls such as `impl<T: SomeTrait> SomeTrait for Option<T> {}`. If you want this impl to work with `Option<Void>`, then you need to be able to implement `SomeTrait` for `Void`. However you also want to tell the compiler that the function is bogus and should not be callable.

We have one such example in [planus](https://github.com/planus-org/planus/blob/7d36b6ab9b36e799a5b6378aed7831cfac8fd06f/crates/planus/src/impls/unit.rs#L7) where we are currently considering disabling this lint.

**Comment by asquared31415** [comment]
As another data point, this is the implementation of every trait on `!` in core, which is useful for being generic.

**Comment by xFrednet** [comment]
I'm reopening this issue, as the linked PR didn't actually resolve it but made the lint allow.by-default instead

**Comment by GrantGryczan** [comment]
This issue's title should be edited to add the `s` at the end of the lint name, so that this issue shows up under the "Related Issues" link of the lint in Clippy's docs.

**Comment by KisaragiEffective** [comment]
another case (minimized):

```rs
use core::convert::Infallible;

#[warn(clippy::uninhabited_references)]
pub fn ignore_infallible(x: &Result<u64, Infallible>) -> u64 {
  match x {
    Ok(v) => v + 1,
    Err(e) => {
      match *e {
      }
    }
  }
}
```

the lint does not like deref:

```
    Checking playground v0.0.1 (/playground)
warning: dereferencing a reference to an uninhabited type is undefined behavior
 --> src/lib.rs:8:13
  |
8 |       match *e {
  |             ^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#uninhabited_references
note: the lint level is defined here
 --> src/lib.rs:3:8
  |
3 | #[warn(clippy::uninhabited_references)]
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.29s
```

**Comment by KisaragiEffective** [comment]
I will look into this once I get some time, I think we agreed that (at least) matching on uninhabited types itself is not UB (but creating reference to those is UB), so that operation must be problematic before it; implies that match should not trigger lint.

**Comment by asquared31415** [comment]
Creating a reference to an uninhabited type is not necessarily UB either, that's currently something that's undecided, and I believe leaning towards "it's valid", but I don't recall, you'd need to go through a bunch of T-opsem meetings to figure out exactly.

**Comment by melody-rs** [comment]
This triggers on `&raw const`, even though miri says its okay? No actual dereference happens here (according to the docs for `std::ptr::addr_of!()` which uses `&raw const`)

```rs
enum Invalid {}

fn main() {
    let v = (0i32, 0i32);
    let ptr = (&raw const v).cast::<(i32, Invalid)>();
    let ptr_to_invalid = unsafe { &raw const (*ptr).0 };
}
```
This is less of an issue for when extern types are eventually stabilized I suppose since they'd replace this sort of code
